### PR TITLE
A4A > Referrals: Implement Client Checkout

### PR DESCRIPTION
**Pull request by yashwin**
Closes https://github.com/Automattic/jetpack-genesis/issues/372

## Proposed Changes

This PR implements client checkout

NOTE: 

- [Design](https://www.figma.com/design/fuufP6VNfZXYmvLsTRWRlG/Referrals?node-id=7600-30663&m=dev)
- **IMPORTANT**: **Please don't forget to do regression testing by logging into the checkout and payment methods page as an agency.**

## Testing Instructions

1. Checkout to the branch locally and start the server.
2. Create a new WPCOM account, but don't sign up for A4A if you have not tested any PR related to the client before.
3. Get the ID(WPCOM ID) from this API that gets called when you log in: 
https://public-api.wordpress.com/rest/v1.1/me and note it down.
4. Login to your agency account > Go to Marketplace > Toggle Refer products > Select product > Go to Checkout > Enter the email ID of the newly created account > Open the network tab > Click the `Request payment from client` button > Get the ID (REFERRAL_ID) from the API response that gets called when the button is clicked.
6. Logout and log in to your client account 
7. Go to /client/payment-methods and delete all the payment methods.
8. Follow the instructions here to get the secret: 34c3e-pb
9. Go to /client/checkout?agency_id={id}&referral_id={id}&secret={secret}
10. Verify that you see the Checkout UI as shown below & verify that the products listed are correct.

Screenshot 2024-06-13 at 11 28 43 AM

11. Click the `Add my payment method` button > Verify that you are directed to the add payment page as shown below:

Screenshot 2024-06-14 at 3 19 06 PM

12. Add a new card & verify that you are redirected to the checkout and it looks like below:

Screenshot 2024-06-14 at 2 41 57 PM

13. Click the `Submit my payment information` and verify that you get a success message and you are redirected to /client/subscriptions.
14. Go to /client/checkout?agency_id={id}&referral_id={id}&secret={secret} again and try submitting it again for the same referral and verify you get the appropriate error message and you are redirected to the subscriptions page.

Screenshot 2024-06-14 at 6 31 35 PM

Other Error/success notices.

Screenshot 2024-06-12 at 5 56 20 PM
Screenshot 2024-06-12 at 6 03 00 PM
Screenshot 2024-06-12 at 6 04 26 PM

15. Please remember to perform regression testing by logging in as an agency on the checkout and payment methods page.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Comment by madebynoam** [comment]
Thank you for the screenshots, @yashwin! 🙌

Leaving a few notes in case you have the bandwidth to tackle them:

1. Can we make some small adjustments to the Payment Methods page? Fyi @keoshi and @jeffgolenski 
    a. Let's change the copy, since the client is not issuing licenses (their agency will):
    **Add a new card** -> **Add your payment method**
    b. Remove the copy about issuing.
    c. Set a max width on the left container to 600px; And space-between its container
    d. Left align the Stripe content and move it up
image

2. Let's refine the copy: "Your payment method has been successfully added." -> "Your payment method was added successfully"

3. Above the Terms text, I see there might be a hidden field that's causing your password manager to show up. Can we fix that?

4. Let's add a space between the terms paragraph and message below it, and can we try to match the sentence widths:
 
image

Otherwise, we're good to go. I see an opportunity for us to have copy style-guide in the future. On my list!

File: client/a8c-for-agencies/sections/marketplace/checkout/style.scss [hunk 1]
```diff
@@ -230,3 +231,40 @@
...
+		font-weight: 600;
+		text-wrap: nowrap;
+	}
+
+	.checkout__summary-notice {
+		margin-block-start: 48px;
+	}
+
+	.checkout__aside-footer {
+		position: absolute;
```
Comment (jkguidaven) [inline client/a8c-for-agencies/sections/marketplace/checkout/style.scss]: Maybe we should avoid using absolute positioning and instead use flex (space-between) due to scalability issues. wdyt?


Here is a checkout with 10 items, causing the footer to overlap.

Screenshot 2024-06-13 at 10 05 15 PM

**Review by jkguidaven — APPROVED** [review]
I have tested the flow according to the test instructions, and it seems to be working as expected. I also performed some regression testing on the purchase flow for normal agency purchases and did not find any issues.

I left one comment, but it's not a major issue, so it shouldn't block the pull request.

Asides the feedbacks from Noam. In terms of the code, it LGTM 👍 nice work!

**Comment by yashwin** [comment]
Thanks for the review, @madebynoam & @jkguidaven!

> a. Let's change the copy, since the client is not issuing licenses (their agency will): Add a new card -> Add your payment method
> b. Remove the copy about issuing.
> c. Set a max width on the left container to 600px; And space-between its container
> d. Left align the Stripe content and move it up

Done: https://github.com/Automattic/wp-calypso/pull/91713/commits/4825a0045b5d380aea9e659a17a6bdb2563aa7df

> Let's refine the copy: "Your payment method has been successfully added." -> "Your payment method was added successfully"

I will update this in another PR as this will need translations.

> Above the Terms text, I see there might be a hidden field that's causing your password manager to show up. Can we fix that?

Some issue with my password manager, I guess. 

> Let's add a space between the terms paragraph and message below it, and can we try to match the sentence widths:

Done, but we cannot match the sentence width since we are forcefully breaking the `Terms & Conditions` link to the next line.

> Maybe we should avoid using absolute positioning and instead use flex (space-between) due to scalability issues. wdyt?

Agree. Updated here:https://github.com/Automattic/wp-calypso/pull/91713/commits/4825a0045b5d380aea9e659a17a6bdb2563aa7df

**Review by madebynoam — APPROVED** [review]
Thanks, @yashwin! Lgtm!

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/14735215

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @yashwin for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/14735215).

**Pull request merged**